File: LC_JAVA/Minimum_no_of_jumps.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

public class Minimum_no_of_jumps {

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t;
        t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            String line = br.readLine();
            String[] tokens = line.split(" ");

            // Create an ArrayList to store the integers
            ArrayList<Integer> array = new ArrayList<>();

            // Parse the tokens into integers and add to the array
            for (String token : tokens) {
                array.add(Integer.parseInt(token));
            }

            int[] arr = new int[array.size()];
            int idx = 0;
            for (int i : array) arr[idx++] = i;

            System.out.println(new Solution().minJumps(arr));
            // System.out.println("~");
        }
    }
}

// } Driver Code Ends


class Solution {
    static int minJumps(int[] arr) {
        // code here
        int destination = arr.length-1;
        int maxreach = 0;
        int windowend = arr[0];
        int jump = 0;
        
        for(int i=0;i<arr.length;i++){
            maxreach = Math.max(maxreach, arr[i] + i);
            if(windowend == i){
                jump++;
                windowend = maxreach;
            }

            if(windowend >= destination){
                return ++jump;
            }
        }

        return -1;        
    }
}

File: LC_CPP/Minimum_no_of_jumps.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// Function to return minimum number of jumps to end of array

class Solution {
  public:
    int minJumps(vector<int>& arr) {
        // code here

        // if(arr[0] == 0){
        //     return -1;
        // }

        int destination = arr.size()-1;
        int maxreach = 0;
        int windowend = arr[0];
        int jump = 0;
        
        for(int i=0;i<arr.size();i++){
            maxreach = max(maxreach, arr[i] + i);
            if(windowend == i){
                jump++;
                windowend = maxreach;
            }

            if(windowend >= destination){
                return ++jump;
            }
        }

        return -1;
        
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        int n, i, j;
        vector<int> arr;
        string ip;
        int number;
        getline(cin, ip);
        stringstream ss(ip);

        while (ss >> number) {
            arr.push_back(number);
        }
        Solution obj;
        cout << obj.minJumps(arr) << endl << "~\n";
    }
    return 0;
}

// } Driver Code Ends